ivy/src/main/scala/sbt/internal/librarymanagement/IvyUtil.scala

```diff
@@ -1,6 +1,50 @@
 package sbt.internal.librarymanagement
 
+import java.io.IOException
+import java.net.{ SocketException, SocketTimeoutException }
+
+import scala.annotation.tailrec
+import scala.util.{ Failure, Success, Try }
+
 private[sbt] object IvyUtil {
   def separate[A, B](l: Seq[Either[A, B]]): (Seq[A], Seq[B]) =
     (l.flatMap(_.left.toOption), l.flatMap(_.right.toOption))
+
+  @tailrec
+  final def retryWithBackoff[T](
+      f: => T,
+      predicate: Throwable => Boolean,
+      maxAttempts: Int,
+      retry: Int = 0
+  ): T = {
+    // Using Try helps in catching NonFatal exceptions only
+    Try {
+      f
+    } match {
+      case Success(value)                                          => value
+      case Failure(e) if predicate(e) && retry < (maxAttempts - 1) =>
+        // max 8s backoff
+        val backoff = math.min(math.pow(2d, retry.toDouble).toLong * 1000L, 8000L)
+        Thread.sleep(backoff)
+        retryWithBackoff(f, predicate, maxAttempts, retry + 1)
+      case Failure(e) => throw e
+    }
+  }
+
+  object TransientNetworkException {
+    private val _r = """.*HTTP response code: (503|429).*""".r
```

Comment:
What about extending the list to:
1. REQUEST_TIMEOUT = 408
2. BAD_GATEWAY = 502
3. GATEWAY_TIMEOUT = 504

Reply:
Looking at Graldle's code it seems to extend this to all 5XX according to the comment - https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifier.java#L27

Reply:
should we implement this? Regex would look something like like: `.*HTTP response code: (5\d{2}|408|429).*`

Reply:
Sure why not?

Reply:
@izharahmd will you do it or should I?

Reply:
@rdesgroppes I have drafted the PR. Adding test cases which i missed in the last PR :)